### ISSUE TITLE
Bug: `filter` callback in `getFirstToken`/`getLastToken` is called with `undefined`

### ISSUE THREAD

**Issue by fasttime** [eslint/eslint#16742]
### Environment

Node version: v19.3.0
npm version: v9.2.0
Local ESLint version: v8.31.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 22.1.0

### What parser are you using?

Default (Espree)

### What did you do?

While working on a custom rule I noticed an edge case where the `filter` callback passed to `sourceCode.getFirstToken()` and `sourceCode.getLastToken()` was being called with an `undefined` argument instead of a `Token` object.

The following is a minimal example I created to illustrate the problem.

```js
const { RuleTester } = require('eslint');

function isCommaToken(token) {
    return token.type === 'Punctuator' && token.value === ',';
}

const rule = {
    meta: {
        docs: { description: 'Forbid commas' },
        messages: { fail: 'commas not allowed' },
    },
    create(context) {
        const sourceCode = context.getSourceCode();
        return {
            Program: node => {
                if (sourceCode.getFirstToken(node, { filter: isCommaToken })) {
                    context.report({ node, messageId: 'fail' });
                }
            }
        };
    }
};

new RuleTester().run(
    'no-commas',
    rule,
    {
        valid:
        [
            '// comment',
        ],
        invalid: [],
    },
);
```

### What did you expect to happen?

The sample test code should run without errors.

### What actually happened?

An error occurs, because the `filter` callback is called with `undefined`:

<pre>
TypeError: Cannot read properties of undefined (reading 'type')
Occurred while linting <input>:1
Rule: "no-commas"
    at isCommaToken (.../no-commas.js:4:18)
    at FilterCursor.moveNext (.../node_modules/eslint/lib/source-code/token-store/filter-cursor.js:37:17)
    at FilterCursor.getOneToken (.../node_modules/eslint/lib/source-code/token-store/cursor.js:49:21)
    at SourceCode.getFirstToken (.../node_modules/eslint/lib/source-code/token-store/index.js:264:11)
    at Program (.../no-commas.js:16:32)
    at ruleErrorHandler (.../node_modules/eslint/lib/linter/linter.js:1115:28)
    at .../node_modules/eslint/lib/linter/safe-emitter.js:45:58
    at Array.forEach (<anonymous>)
    at Object.emit (.../node_modules/eslint/lib/linter/safe-emitter.js:45:38)
    at NodeEventGenerator.applySelector (.../node_modules/eslint/lib/linter/node-event-generator.js:297:26) {
  ruleId: 'no-commas',
  currentNode: Node {
    type: 'Program',
    loc: SourceLocation {
      start: Position { line: 1, column: 0 },
      end: Position { line: 1, column: 10 }
    },
    range: [ 0, 10 ],
    body: [],
    sourceType: 'script',
    comments: [
      {
        type: 'Line',
        value: ' comment',
        range: [ 0, 10 ],
        loc: {
          start: Position { line: 1, column: 0 },
          end: Position { line: 1, column: 10 }
        }
      }
    ],
    tokens: [],
    parent: null
  }
}
</pre>

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

The problem occurs when a file being linted <s>is empty or</s> contains only comments, which is not extremely uncommon in practice.

I checked the usages of `getFirstToken` and `getLastToken` in the code base and it seems that none of the core rule is affected by this problem, but since these methods are part of the public API, I am wondering if this issue should be fixed. If not, it would be great to have it documented that the callback argument could be undefined sometimes, although this does not seem to be the intended behavior.

**Comment by fasttime** [comment]
@mdjermanovic you're right! This does not happen with an empty file. Sorry for the confusion, not sure what I was doing when I tested it.

I can see that [`ForwardTokenCursor#indexEnd`](https://github.com/eslint/eslint/blob/8108f49f9fa0c2de80b3b66c847551beff585951/lib/source-code/token-store/forward-token-cursor.js#L35) is set 0 when the source file contains only comments. It should probably be -1 like when the source is empty. Similarly, [`BackwardTokenCursor#index`](https://github.com/eslint/eslint/blob/8108f49f9fa0c2de80b3b66c847551beff585951/lib/source-code/token-store/backward-token-cursor.js#L34) should start at -1. I'd have to look deeper into this to know for sure.

**Comment by fasttime** [comment]
After digging more into this issue, I managed to came up with a possible fix.
I believe the error is caused by a misassumption that a node should always contain one or more tokens, which doesn't necessarily hold true for some types of node, like `Program`.
My fix consists in treating a missing token (when there are no tokens) in a `ForwardTokenCursor` or `BackwardTokenCursor` like an immediate match on the current index, rather than on the next or previous index.
This prevents the cursor from setting its initial token index out of the bounds of the `tokens` array.
And the ultimate effect is that the `TokenStore` methods no longer trigger a bogus call to the `filter` callback and return `null` instead of `undefined`.
I drafted a PR to show the changes.
